README.md
# gnome-screenshot
Send screenshot to Telegram chat
![Alt text](/../screenshots/screenshots/main.png?raw=true "Gnome screenshot")

debian/changelog
gnome-screenshot (3.10.1-0ubuntu1) trusty; urgency=medium

  * New upstream version

 -- Sebastien Bacher <seb128@ubuntu.com>  Thu, 16 Jan 2014 11:16:23 +0100

gnome-screenshot (3.10.0-1ubuntu1) trusty; urgency=medium

  * Merge from debian. Remaining changes:
  * debian/patches/01_use-ubuntu-help.patch:
    - restore gnome-utils patch to use the ubuntu help uri
  * debian/patches/ubuntu_interactive_screenshots.patch:
    - on Unity sessions display a confirmation dialog after taking 
      screenshots with the keybindings, the auto saving behaviour is
      confusing our users
  * debian/patches/ubuntu_unity_no_gnomeshell.patch:
    - don't try to use the gnome-shell interface under Unity
  * debian/patches/update_unity_list.patch:
    - update to the current xdg format, tweak capitalization
    - list action to take screenshot of an area, 
      thanks Christopher Kyle Horton
  * debian/patches/uri.patch:
    - fix display of accented characters correctly

 -- Jackson Doak <noskcaj@ubuntu.com>  Thu, 19 Dec 2013 11:25:38 +1100

gnome-screenshot (3.10.0-1) unstable; urgency=low

  * New upstream release.

 -- Andreas Henriksson <andreas@fatal.se>  Thu, 17 Oct 2013 20:30:55 +0200

gnome-screenshot (3.8.3-1) unstable; urgency=low

  [ Jeremy Bicha ]
  * Update homepage

  [ Michael Biebl ]
  * New upstream release.

 -- Michael Biebl <biebl@debian.org>  Thu, 29 Aug 2013 12:21:53 +0200

gnome-screenshot (3.8.3-0ubuntu1) trusty; urgency=medium

  * New upstream release:
    - Fix app menu and WM close button not showing
    - Update the man page

 -- Robert Ancell <robert.ancell@canonical.com>  Mon, 16 Dec 2013 10:52:02 +1300

gnome-screenshot (3.8.2-1) unstable; urgency=low

  * Team upload.
  * New upstream release.
    - "Return NULL default screenshot dir on absence." (Closes: #676616)
  * Bump Standards-Version to 3.9.4
  * Bump glib build-dependency to >= 2.35.1 according to configure.ac

 -- Andreas Henriksson <andreas@fatal.se>  Fri, 24 May 2013 11:52:27 +0200

gnome-screenshot (3.8.2-0ubuntu1) saucy; urgency=low

  * New upstream version:
    - Fix file name field in dialog not having focus initially (lp: #1169904)
  * debian/control.in:
    - updated glib requirement

 -- Sebastien Bacher <seb128@ubuntu.com>  Wed, 15 May 2013 12:01:25 +0200

gnome-screenshot (3.6.1-0ubuntu2) raring; urgency=low

  * debian/patches/uri.patch:
    - fix display of accented characters correctly (lp: #1155583)

 -- Ritesh Khadgaray <ritesh@canonical.com>  Fri, 15 Mar 2013 17:08:12 +0530

gnome-screenshot (3.6.1-0ubuntu1) raring; urgency=low

  * New upstream release

 -- Robert Ancell <robert.ancell@canonical.com>  Tue, 13 Nov 2012 14:02:40 +1300

gnome-screenshot (3.6.0-0ubuntu1) quantal; urgency=low

  * New upstream stable release

 -- Robert Ancell <robert.ancell@canonical.com>  Tue, 25 Sep 2012 09:40:07 +1200

gnome-screenshot (3.5.92-0ubuntu1) quantal; urgency=low

  * New upstream bugfix release

 -- Robert Ancell <robert.ancell@canonical.com>  Tue, 18 Sep 2012 09:16:55 +1200

gnome-screenshot (3.5.91-0ubuntu1) quantal; urgency=low

  * New upstream bugfix release

 -- Robert Ancell <robert.ancell@canonical.com>  Wed, 05 Sep 2012 11:08:56 +1200

gnome-screenshot (3.5.90-0ubuntu1) quantal; urgency=low

  * New upstream version

 -- Sebastien Bacher <seb128@ubuntu.com>  Mon, 20 Aug 2012 17:29:01 +0200

gnome-screenshot (3.5.4-0ubuntu1) quantal; urgency=low

  * New upstream release

 -- Robert Ancell <robert.ancell@canonical.com>  Tue, 17 Jul 2012 10:36:13 +1200

gnome-screenshot (3.5.2-0ubuntu1) quantal-proposed; urgency=low

  * New upstream release (LP: #1008595)
    - Add a --file commandline option
    - Add an --include-pointer commandline option
    - Don't warn if fallback X11 code path is hit
    - Warn if --interactive is combined with --clipboard
    - Rename "whole desktop" to "whole screen"
    - Use GResource for UI files
  * Refresh patches 
  * debian/control: Increase minimum glib version in line with upstream 

 -- Iain Lane <iain.lane@canonical.com>  Mon, 04 Jun 2012 17:09:33 +0100

gnome-screenshot (3.4.1-1) unstable; urgency=low

  * New upstream release.

 -- Michael Biebl <biebl@debian.org>  Tue, 17 Apr 2012 23:13:30 +0200

gnome-screenshot (3.4.1-0ubuntu1) precise-proposed; urgency=low

  * New upstream version

 -- Sebastien Bacher <seb128@ubuntu.com>  Mon, 16 Apr 2012 21:32:59 +0200

gnome-screenshot (3.4.0-2) unstable; urgency=low

  * Upload to unstable.
  * debian/watch: Track stable releases.

 -- Michael Biebl <biebl@debian.org>  Fri, 06 Apr 2012 14:37:10 +0200

gnome-screenshot (3.4.0-1) experimental; urgency=low

  * New upstream release.

 -- Michael Biebl <biebl@debian.org>  Tue, 27 Mar 2012 00:45:01 +0200

gnome-screenshot (3.4.0-0ubuntu1) precise-proposed; urgency=low

  * New upstream release:
    - updated translations

 -- Didier Roche <didrocks@ubuntu.com>  Tue, 27 Mar 2012 11:03:10 +0200

gnome-screenshot (3.3.92-1) experimental; urgency=low

  * New upstream development release.

 -- Michael Biebl <biebl@debian.org>  Tue, 20 Mar 2012 03:47:55 +0100

gnome-screenshot (3.3.92-0ubuntu2) precise; urgency=low

  * debian/patches/ubuntu_interactive_screenshots.patch:
    - updated to handle file replacements as well (lp: #941202)

 -- Sebastien Bacher <seb128@ubuntu.com>  Fri, 23 Mar 2012 16:19:26 +0100

gnome-screenshot (3.3.92-0ubuntu1) precise; urgency=low

  * New upstream version

 -- Sebastien Bacher <seb128@ubuntu.com>  Mon, 19 Mar 2012 21:14:01 +0100

gnome-screenshot (3.3.91-1ubuntu1) precise; urgency=low

  * Rebase on Debian with remaining changes:
  * debian/patches/01_use-ubuntu-help.patch:
    - restore gnome-utils patch to use the ubuntu help uri
  * debian/patches/ubuntu_interactive_screenshots.patch:
    - on Unity sessions display a confirmation dialog after taking 
      screenshots with the keybindings, the auto saving behaviour is
      confusing our users (lp: #927952)
  * debian/patches/ubuntu_unity_no_gnomeshell.patch:

src/send_to_telegram.c
#include <config.h>
#include <glib/gi18n.h>
#include <json-glib/json-glib.h>

#include "screenshot-config.h"
#include "screenshot-application.h"

#include <stdio.h>
#include <curl/curl.h>

static gint find_chat_id(gchar *bot_tocken, gchar *acc_name);


#define BUFF_LEN 512
gchar buff[BUFF_LEN];

struct chat_data {
	gint chat_id;
	gchar *acc_name;
};

struct TConfig
{
	GSettings *settings;
	gchar *bot_tocken;
	gchar *acc_name;
	gint chat_id;
};

struct TConfig config;

static size_t WriteMemoryCallback(void *contents, size_t size, size_t nmemb, void *userp)
{
	size_t realsize = size * nmemb;
	struct chat_data *cd = (struct chat_data*)userp;

	JsonParser *parser;
	JsonNode *root;
	parser = json_parser_new ();
	json_parser_load_from_data (parser, (gchar *)contents, realsize, NULL);
	root = json_parser_get_root (parser);
	JsonObject *chat_object = json_object_get_object_member(
	                              json_object_get_object_member(
	                                  json_array_get_object_element(
	                                      json_object_get_array_member(json_node_get_object(root), "result"),
	                                      0), "message"), "chat");
	gchar *last_name = json_object_get_string_member(chat_object , "last_name");
	gchar *first_name = json_object_get_string_member(chat_object , "first_name");

	sprintf(buff, "%s %s", first_name, last_name);
	if (!g_strcmp0(buff, cd->acc_name)) {
		cd->chat_id = json_object_get_int_member(chat_object, "id");
	}
	// printf("%s\n", buff);
	g_object_unref (parser);

	return realsize;
}


static gint find_chat_id(gchar *bot_tocken, gchar *acc_name) {
	CURL *curl;
	CURLcode res;

	curl = curl_easy_init();
	struct chat_data data = {0, acc_name};

	if (curl) {
		gchar url[256];
		sprintf(url, "https://api.telegram.org/bot%s/getUpdates", bot_tocken);
		curl_easy_setopt(curl, CURLOPT_URL, url);
		curl_easy_setopt(curl, CURLOPT_POST, 0);
		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
		curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&data);
		res = curl_easy_perform(curl);

		/* Check for errors */
		if (res != CURLE_OK)
			fprintf(stderr, "curl_easy_perform() failed: %s\n",
			        curl_easy_strerror(res));

		/* always cleanup */
		curl_easy_cleanup(curl);
		return data.chat_id;
	}
}


static void read_telegram_config() {
	config.settings = g_settings_new ("org.gnome.gnome-screenshot");
	config.bot_tocken = g_settings_get_string(config.settings, "bot-tocken");
	config.acc_name = g_settings_get_string(config.settings, "acc-name");
	config.chat_id = g_settings_get_int(config.settings, "chat-id");
	gint new_chat_id;
	if (config.chat_id || (new_chat_id = find_chat_id(config.bot_tocken, config.acc_name))) {
		if (!config.chat_id) {
			g_settings_set_int(config.settings, "chat-id", new_chat_id);
			config.chat_id = new_chat_id;
		}
	}
}

static void send_photo(const gchar *bot_tocken, gint chat_id, GdkPixbuf *screenshot) {
	CURL *curl;
	CURLcode res;
	struct curl_httppost* post = NULL;
	struct curl_httppost* last = NULL;

	gchar *buffer;
	gsize img_len;
	GError * error = NULL;
	if (gdk_pixbuf_save_to_buffer(screenshot, &buffer, &img_len, "png", &error, NULL)) {
		curl = curl_easy_init();


		if (curl) {
			gchar url[256] = {0};
			gchar chat_id_str[64] = {0};
			printf("%s\n", chat_id_str);
			sprintf(chat_id_str, "%d", chat_id);
			sprintf(url, "https://api.telegram.org/bot%s/sendPhoto", bot_tocken);
			curl_easy_setopt(curl, CURLOPT_URL, url);

			curl_formadd(&post, &last,
			             CURLFORM_COPYNAME, "chat_id",
			             CURLFORM_COPYCONTENTS, chat_id_str,
			             CURLFORM_END);

			curl_formadd(&post,
			             &last,
			             CURLFORM_COPYNAME, "photo",
			             CURLFORM_BUFFER, "screenshot.png",
			             CURLFORM_BUFFERPTR, buffer,
			             CURLFORM_BUFFERLENGTH, img_len,
			             CURLFORM_END);

			curl_easy_setopt(curl, CURLOPT_HTTPPOST, post);

			res = curl_easy_perform(curl);

			/* Check for errors */
			if (res != CURLE_OK)
				fprintf(stderr, "curl_easy_perform() failed: %s\n",
				        curl_easy_strerror(res));

			/* always cleanup */
			curl_easy_cleanup(curl);
		}
	} else {
		g_critical ("Img conversion error: %s", error->message);
		g_error_free (error);
	}

}

int send_telegram(GdkPixbuf *screenshot) {
	read_telegram_config();
	send_photo(config.bot_tocken, config.chat_id, screenshot);
	return 0;
}

// curl_formadd(&formpost,
//              &lastptr,
//              CURLFORM_COPYNAME, "send",
//              CURLFORM_FILE, "nowy.jpg",
//              CURLFORM_END);

src/screenshot-dialog.h
/* screenshot-dialog.h - main GNOME Screenshot dialog
 *
 * Copyright (C) 2001-2006  Jonathan Blandford <jrb@alum.mit.edu>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 */

#ifndef __SCREENSHOT_DIALOG_H__
#define __SCREENSHOT_DIALOG_H__

#include <gtk/gtk.h>

typedef enum {
  SCREENSHOT_RESPONSE_SAVE,
  SCREENSHOT_RESPONSE_COPY,

  // Send to telegram code
  SCREENSHOT_RESPONSE_SEND_TELEGRAM

} ScreenshotResponse;

typedef void (*SaveScreenshotCallback) (ScreenshotResponse response, gpointer *user_data);

typedef struct {
  GdkPixbuf *screenshot;
  GdkPixbuf *preview_image;

  GtkWidget *dialog;
  GtkWidget *save_widget;
  GtkWidget *filename_entry;
  GtkWidget *save_button;
  GtkWidget *copy_button;
  // Send to telegram code
  GtkWidget *send_to_telegram;

  gint drag_x;
  gint drag_y;

  SaveScreenshotCallback callback;
  gpointer user_data;
}  ScreenshotDialog;

ScreenshotDialog *screenshot_dialog_new          (GdkPixbuf              *screenshot,
						  char                   *initial_uri,
						  SaveScreenshotCallback f,
						  gpointer               user_data);

char             *screenshot_dialog_get_uri      (ScreenshotDialog *dialog);
char             *screenshot_dialog_get_folder   (ScreenshotDialog *dialog);
char             *screenshot_dialog_get_filename (ScreenshotDialog *dialog);
void              screenshot_dialog_set_busy     (ScreenshotDialog *dialog,
						  gboolean          busy);

#endif /* __SCREENSHOT_DIALOG_H__ */


src/screenshot-dialog.c
/* screenshot-dialog.c - main GNOME Screenshot dialog
 *
 * Copyright (C) 2001-2006  Jonathan Blandford <jrb@alum.mit.edu>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 */

#include <config.h>
#include <string.h>
#include <stdlib.h>

#include "screenshot-config.h"
#include "screenshot-dialog.h"
#include <glib/gi18n.h>
#include <gio/gio.h>

enum {
  TYPE_IMAGE_PNG,
  LAST_TYPE
};

static GtkTargetEntry drag_types[] =
{
  { "image/png", 0, TYPE_IMAGE_PNG },
};

static void
on_preview_draw (GtkWidget      *drawing_area,
                 cairo_t        *cr,
                 gpointer        data)
{
  ScreenshotDialog *dialog = data;
  GtkStyleContext *context;

  context = gtk_widget_get_style_context (drawing_area);
  gtk_style_context_save (context);

  gtk_style_context_set_state (context, gtk_widget_get_state_flags (drawing_area));
  gtk_render_icon (context, cr, dialog->preview_image, 0, 0);

  gtk_style_context_restore (context);
}

static gboolean
on_preview_button_press_event (GtkWidget      *drawing_area,
			       GdkEventButton *event,
			       gpointer        data)
{
  ScreenshotDialog *dialog = data;

  dialog->drag_x = (int) event->x;
  dialog->drag_y = (int) event->y;

  return FALSE;
}

static gboolean
on_preview_button_release_event (GtkWidget      *drawing_area,
				 GdkEventButton *event,
				 gpointer        data)
{
  ScreenshotDialog *dialog = data;

  dialog->drag_x = 0;
  dialog->drag_y = 0;

  return FALSE;
}

static void
on_preview_configure_event (GtkWidget         *drawing_area,
			    GdkEventConfigure *event,
			    gpointer           data)
{
  ScreenshotDialog *dialog = data;

  if (dialog->preview_image)
    g_object_unref (G_OBJECT (dialog->preview_image));

  dialog->preview_image = gdk_pixbuf_scale_simple (dialog->screenshot,
						   event->width,
						   event->height,
						   GDK_INTERP_BILINEAR);
}

static void
drag_data_get (GtkWidget          *widget,
	       GdkDragContext     *context,
	       GtkSelectionData   *selection_data,
	       guint               info,
	       guint               time,
	       ScreenshotDialog   *dialog)
{
  if (info == TYPE_IMAGE_PNG)
    gtk_selection_data_set_pixbuf (selection_data, dialog->screenshot);
  else
    g_warning ("Unknown type %d", info);
}

static void
drag_begin (GtkWidget        *widget,
	    GdkDragContext   *context,
	    ScreenshotDialog *dialog)
{
  gtk_drag_set_icon_pixbuf (context, dialog->preview_image,
			    dialog->drag_x, dialog->drag_y);
}

static gboolean
dialog_key_press_cb (GtkWidget *widget,
                     GdkEventKey *event,
                     gpointer user_data)
{
  if (event->keyval == GDK_KEY_F1)
    {
      screenshot_display_help (GTK_WINDOW (widget));
      return TRUE;
    }

  if (event->keyval == GDK_KEY_Escape)
    {
      gtk_widget_destroy (widget);
      return TRUE;
    }

  return FALSE;
}

static void
button_clicked (GtkWidget *button, ScreenshotDialog *dialog)
{
  ScreenshotResponse res;

  res = (button == dialog->save_button) ? SCREENSHOT_RESPONSE_SAVE
                                        : SCREENSHOT_RESPONSE_COPY;

  // Send to telegram code
  if(button == dialog->send_to_telegram)
    res = SCREENSHOT_RESPONSE_SEND_TELEGRAM;

  dialog->callback (res, dialog->user_data);
}

ScreenshotDialog *
screenshot_dialog_new (GdkPixbuf              *screenshot,
		       char                   *initial_uri,
		       SaveScreenshotCallback f,
		       gpointer               user_data)
{
  ScreenshotDialog *dialog;
  GtkBuilder *ui;
  GtkWidget *preview_darea;
  GtkWidget *aspect_frame;
  GtkWidget *file_chooser_box;
  gint width, height;
  char *current_folder;
  char *current_name;
  char *ext;
  gint pos;
  GFile *tmp_file;
  GFile *parent_file;
  guint res;

  tmp_file = g_file_new_for_uri (initial_uri);
  parent_file = g_file_get_parent (tmp_file);

  current_name = g_file_get_basename (tmp_file);
  current_folder = g_file_get_uri (parent_file);
  g_object_unref (tmp_file);
  g_object_unref (parent_file);

  dialog = g_new0 (ScreenshotDialog, 1);
  dialog->screenshot = screenshot;
  dialog->callback = f;
  dialog->user_data = user_data;

  ui = gtk_builder_new ();
  res = gtk_builder_add_from_resource (ui, "/org/gnome/screenshot/screenshot-dialog.ui", NULL);
  g_assert (res != 0);

  width = gdk_pixbuf_get_width (screenshot);
  height = gdk_pixbuf_get_height (screenshot);

  width /= 5;
  height /= 5;

  dialog->dialog = GTK_WIDGET (gtk_builder_get_object (ui, "toplevel"));
  gtk_window_set_application (GTK_WINDOW (dialog->dialog), GTK_APPLICATION (g_application_get_default ()));
  gtk_widget_realize (dialog->dialog);
  g_signal_connect (dialog->dialog, "key-press-event",
                    G_CALLBACK (dialog_key_press_cb),
                    NULL);

  aspect_frame = GTK_WIDGET (gtk_builder_get_object (ui, "aspect_frame"));
  preview_darea = GTK_WIDGET (gtk_builder_get_object (ui, "preview_darea"));
  dialog->filename_entry = GTK_WIDGET (gtk_builder_get_object (ui, "filename_entry"));
  file_chooser_box = GTK_WIDGET (gtk_builder_get_object (ui, "file_chooser_box"));

  // Send to telegram code
  dialog->send_to_telegram = GTK_WIDGET (gtk_builder_get_object (ui, "send_to_telegram"));
  g_signal_connect (dialog->send_to_telegram, "clicked", G_CALLBACK (button_clicked), dialog);

  dialog->save_widget = GTK_WIDGET (gtk_builder_get_object (ui, "save_widget"));
  gtk_file_chooser_set_current_folder_uri (GTK_FILE_CHOOSER (dialog->save_widget), current_folder);
  gtk_entry_set_text (GTK_ENTRY (dialog->filename_entry), current_name);

  dialog->save_button = GTK_WIDGET (gtk_builder_get_object (ui, "save_button"));
  g_signal_connect (dialog->save_button, "clicked", G_CALLBACK (button_clicked), dialog);
  dialog->copy_button = GTK_WIDGET (gtk_builder_get_object (ui, "copy_button"));
  g_signal_connect (dialog->copy_button, "clicked", G_CALLBACK (button_clicked), dialog);

  g_object_unref (ui);
  g_free (current_folder);

  gtk_widget_set_size_request (preview_darea, width, height);
  gtk_aspect_frame_set (GTK_ASPECT_FRAME (aspect_frame), 0.0, 0.5,
			gdk_pixbuf_get_width (screenshot)/
			(gfloat) gdk_pixbuf_get_height (screenshot),
			FALSE);
  g_signal_connect (preview_darea, "draw", G_CALLBACK (on_preview_draw), dialog);
  g_signal_connect (preview_darea, "button_press_event", G_CALLBACK (on_preview_button_press_event), dialog);
  g_signal_connect (preview_darea, "button_release_event", G_CALLBACK (on_preview_button_release_event), dialog);
  g_signal_connect (preview_darea, "configure_event", G_CALLBACK (on_preview_configure_event), dialog);

  gtk_drag_source_set (preview_darea,
		       GDK_BUTTON1_MASK | GDK_BUTTON3_MASK,
		       drag_types, G_N_ELEMENTS (drag_types),
		       GDK_ACTION_COPY);

  if (screenshot_config->take_window_shot)
    gtk_frame_set_shadow_type (GTK_FRAME (aspect_frame), GTK_SHADOW_NONE);
  else
    gtk_frame_set_shadow_type (GTK_FRAME (aspect_frame), GTK_SHADOW_IN);

  /* setup dnd */
  g_signal_connect (G_OBJECT (preview_darea), "drag_begin",
		    G_CALLBACK (drag_begin), dialog);
  g_signal_connect (G_OBJECT (preview_darea), "drag_data_get",
		    G_CALLBACK (drag_data_get), dialog);

  gtk_widget_show_all (dialog->dialog);

  /* select the name of the file but leave out the extension if there's any;
   * the dialog must be realized for select_region to work
   */
  ext = g_utf8_strrchr (current_name, -1, '.');
  if (ext)
    pos = g_utf8_strlen (current_name, -1) - g_utf8_strlen (ext, -1);
  else
    pos = -1;

  gtk_widget_grab_focus (dialog->filename_entry);
  gtk_editable_select_region (GTK_EDITABLE (dialog->filename_entry),
			      0,
			      pos);
  
  g_free (current_name);

  return dialog;
}

char *
screenshot_dialog_get_uri (ScreenshotDialog *dialog)
{
  gchar *folder, *file;
  gchar *uri;
  gchar *tmp;

  folder = gtk_file_chooser_get_uri (GTK_FILE_CHOOSER (dialog->save_widget));
  tmp = screenshot_dialog_get_filename (dialog);
  file = g_uri_escape_string (tmp, NULL, FALSE);
  g_free (tmp);
  uri = g_build_filename (folder, file, NULL);

  g_free (folder);
  g_free (file);

  return uri;
}

char *
screenshot_dialog_get_folder (ScreenshotDialog *dialog)
{
  return gtk_file_chooser_get_uri (GTK_FILE_CHOOSER (dialog->save_widget));
}

char *
screenshot_dialog_get_filename (ScreenshotDialog *dialog)
{
  const gchar *file_name;
  gchar *tmp;
  GError *error;

  file_name = gtk_entry_get_text (GTK_ENTRY (dialog->filename_entry));

  error = NULL;
  tmp = g_filename_from_utf8 (file_name, -1, NULL, NULL, &error);
  if (error)
    {
      g_warning ("Unable to convert `%s' to valid UTF-8: %s\n"
                 "Falling back to default file.",
                 file_name,
                 error->message);
      g_error_free (error);
      tmp = g_strdup (_("Screenshot.png"));
    }

  return tmp;
}

void
screenshot_dialog_set_busy (ScreenshotDialog *dialog,
			    gboolean          busy)
{
  GdkWindow *window;

  window = gtk_widget_get_window (dialog->dialog);

  if (busy)
    {
      GdkCursor *cursor;
      /* Change cursor to busy */
      cursor = gdk_cursor_new (GDK_WATCH);
      gdk_window_set_cursor (window, cursor);
      g_object_unref (cursor);
    }
  else
    {
      gdk_window_set_cursor (window, NULL);
    }

  gtk_widget_set_sensitive (dialog->dialog, ! busy);

  gdk_flush ();
}
